README.md
# kapture

[![Latest Release](https://img.shields.io/github/v/tag/kalisio/kapture?sort=semver&label=latest)](https://github.com/kalisio/kapture/releases)
[![Build Status](https://app.travis-ci.com/kalisio/kapture.svg?branch=master)](https://app.travis-ci.com/kalisio/kapture)

**Capture Kano Snapshots as a Service**

**kapture** is a lightweight service that let you take screenshots of [Kano](https://kalisio.github.io/kano/). 

**kapture** relies on [Puppeteer](https://github.com/puppeteer/puppeteer) and [Express](https://expressjs.com/fr/)

## API

### capture (POST)

Request a capture with the following query parameters.

The body of the request must conform a **JSON** object with the following properties: 

| Property | Description | Defaults`|
| --- | --- | -- |
| `layers` | specifies the layers to display | `[]` | 
| `activity` | specifies the `map` or `globe` view | `map` |
| `size` | specifies the capture size | `{ "width": 1024, "height": 768 }` |
| `delay` | specified the waiting delay before capturing the screenshot (in milliseconds) | `1000` |

The `layers` property must conform the following JSON schema: 

```json
"layers": {
  "type" : "array",
  "items": {
    "type": "string"
  }
}
```

The items must conform the [kano](https://kalisio.github.io/kano/) formalism, i.e. `Layers.MY_LAYER`. In addition and for backward compatibility, you can specify the layer name using kebab case: `layers-my-layer`, with or without the `layers-` prefix. 

The `activity` property must conform the following JSON schema: 

```json
"activity": {
  "type": "string",
  "pattern": "map|globe"
}
```

The `size` property must conform the following JSON schema: 

```json
"size": {
  "type": "object",
  "properties": {
    "width": {
      "type": "integer",
      "default": 1024
    },
    "height": {
      "type": "integer",
      "default": 768
    }
}
```

The `delay` property must conform the following JSON schema: 

```json
"activity": {
  "type": "number",
  "default": "2000"
}
```

In addition the body can conform a [GeoJSON](https://datatracker.ietf.org/doc/html/rfc7946) object. The described features will be rendered as an overlay.

Here is a complete example:

```json
{
  "layers": ["Layers.IMAGERY"],
  "type": "FeatureCollection",
  "features": [
    { 
      "type": "Feature", 
      "geometry": { 
        "type": "Point", 
        "coordinates": [3, 42.5]
      },
      "properties": { "fill-color": "#AAAAAA" } 
    },
    { 
      "type": "Feature", 
      "geometry": { 
        "type": "LineString", 
        "coordinates": [ [3, 42], [4, 43], [5,42], [6, 43]] 
      },
      "properties": { "fill-color": "#AAAAAA" } 
    },
    { 
      "type": "Feature", 
      "geometry": { 
        "type": "Polygon", 
        "coordinates": [ [ [0, 42], [1, 42], [1, 43], [0, 43], [0, 42] ] ] 
      }, 
      "properties": { "fill-color": "#AAAAAA" } 
    }
  ]
}
```

And the response looks like:

![response](./assets/response.png)

### healthcheck (GET)

Check for the health of the service

## Configuring

Here are the environment variables you can use to customize the service:

| Variable  | Description | Defaults |
|-----------| ------------| ------------|
| `KANO_URL` | The **Kano** url | - |
| `KANO_JWT` | The **Kano** token to get connected | * |
| `BODY_LIMIT` | The size limit of the request body | `100kb` |
| `DELAY` | The waiting delay before capturing the screen (in milliseconds) | '1000' |
| `DEBUG` | The namespaces to enable debug output. Set it to `kapture:*` to enable full debug output. | 

## Building

### Manual build 

You can build the image with the following command:

```bash
docker build -t <your-image-name> .
```

### Automatic build using Travis CI

This project is configured to use **Travis** to build and push the image on the [Kalisio's Docker Hub](https://hub.docker.com/u/kalisio/).
The built image is tagged using the `version` property in the `package.json` file.

To enable Travis to do the job, you must define the following variable in the corresponding Travis project:

| Variable  | Description |
|-----------| ------------|
| `DOCKER_USER` | your username |
| `DOCKER_PASSWORD` | your password |

## Deploying

This image is designed to be deployed using the [Kargo](https://kalisio.github.io/kargo/) project.

Check out the [compose file](https://github.com/kalisio/kargo/blob/master/deploy/kontrol.yml) to have an overview on how the container is deployed.

## Testing

To test the service you need to set the following environment variables:

| Variable  | Description | Defaults |
|-----------| ------------| ------------|
| `KAPTURE_URL` | The **Kapture** url. | `http://localhost:3000` |
| `KAPTURE_JWT` | The **Kapture** bearen token to pass a gateway if needed. It is set using `Auhtorization` header. | - |

To run the tests, use the subcommand `test`: 

```bash
yarn test
```

## Contributing

Please read the [Contributing file](./.github/CONTRIBUTING.md) for details on our code of conduct, and the process for submitting pull requests to us.

## Authors

This project is sponsored by 

![Kalisio](https://s3.eu-central-1.amazonaws.com/kalisioscope/kalisio/kalisio-logo-black-256x84.png)

## License

This project is licensed under the MIT License - see the [license file](./LICENSE.md) for details

CHANGELOG.md
Please refer to GitHub issues for changelog starting from v1.1.0: https://github.com/kalisio/kapture/issues

## 1.0.0 (2022-01-06)

##### New Features

*  Ensure compatibility with old layer identification [close [#9](https://github.com/kalisio/kapture/pull/9)] ([eddd900d](https://github.com/kalisio/kapture/commit/eddd900d7e90914d0f6aea17b66716a85c0786ff))
*  Enhance body validation (close [#8](https://github.com/kalisio/kapture/pull/8)) ([87acbe65](https://github.com/kalisio/kapture/commit/87acbe65d31bd0737306c37dd1a138f1f11bbe6a))
*  Take into account the source activity (map or globe) (close [#6](https://github.com/kalisio/kapture/pull/6)) ([75bd0c55](https://github.com/kalisio/kapture/commit/75bd0c55738acddd5046815fb1f51e889cf413e3))
*  Kapture must take into account the current time (close [#3](https://github.com/kalisio/kapture/pull/3)) wip: When capturing from Kano canvas layers are not well handled [#5](https://github.com/kalisio/kapture/pull/5) ([525598eb](https://github.com/kalisio/kapture/commit/525598ebdb6f7f903a6446a25eff12f605d83276))
*  Check for body size limitation when requesting the capture endpoint (close [#1](https://github.com/kalisio/kapture/pull/1)) ([a160b3c1](https://github.com/kalisio/kapture/commit/a160b3c1ebfc8902fd08063f56d5a4ff44767ff4))
*  Enhance GeoJson support (fix [#4](https://github.com/kalisio/kapture/pull/4)) ([5a3c7de1](https://github.com/kalisio/kapture/commit/5a3c7de1ede026d533277168902642a0d8f243ad))
*  enhance layer id handler ([0e76dc74](https://github.com/kalisio/kapture/commit/0e76dc7476678bdc5518a6048ec6cc84a6a1e524))
*  handle bbox in geojson ([05720799](https://github.com/kalisio/kapture/commit/05720799e39584c99236c31088c36b69f40abc43))

##### Bug Fixes

*  category selector ([1600cd34](https://github.com/kalisio/kapture/commit/1600cd3482cb9035b01aee23e1c4911b1146c8c0))
*  getLayerCategoryId call ([383888e8](https://github.com/kalisio/kapture/commit/383888e842cf029cb55eef670ae62dcd212d10d0))
*  geojsonhnt errors handler is not correct ([613541d1](https://github.com/kalisio/kapture/commit/613541d15fc838a6917da3dcda1a8bc79e38d9ba))





index.js
#!/usr/bin/env node
import _ from 'lodash'
import cors from 'cors'
import express from 'express'
import { Buffer } from 'buffer'
import { capture } from './src/capture.js'
import { validateGeoJson } from './src/utils.geojson.js'

const port = process.env.PORT || 3000
const bodyLimit = process.env.BODY_LIMIT || '100kb'
const delay = process.env.DELAY || 1000
const kanoUrl = process.env.KANO_URL
const kanoJwt = process.env.KANO_JWT

// activity validator middleware
const activityValidator = function (req, res, next) {
  const activity = _.get(req.body, 'activity')
  if (activity) {
    if (!_.includes(['map', 'globe'], activity)) res.status(404).json({ message: 'Invdalid \"activity\" property' })
    else next()
  } else {
    next()
  }
}

// layers validator middleware
const layersValidator = function (req, res, next) {
  const layers = _.get(req.body, 'layers') 
  if (layers) {
    if (!_.isArray(layers)) res.status(404).json({ message: 'Invdalid \"layers\" property' })
    else next()
  } else {
    next()
  }
}

// size validator
const sizeValidator = function (req, res, next) {
  const width = _.get(req.body, 'size.width')
  const height = _.get(req.body, 'size.height')
  if (width || height) {
    if (width < 256 || width > 4000 || height < 256 || height > 4000) res.status(404).json({ message: 'Invdalid \"size\" property' })
    else next()
  } else {
    next()
  }
}

// features validator middleware
const geoJsonValidator = function (req, res, next) {
  if (req.body.type === 'FeatureCollection' || req.body.type === 'Feature') {
    const errors = validateGeoJson(req.body)
    if (errors.length > 0) res.status(422).json({ message: 'Invdalid \"GeoJSON\"', errors })
    else next()
  } else {
    next()
  }
}

// Initialize express app
const app = express()
app.use(cors()) // enable cors
app.use(express.urlencoded({ limit: bodyLimit, extended: true }))
app.use(express.json({ limit: bodyLimit }))

// Capture 
app.post('/capture', [activityValidator, layersValidator, sizeValidator, geoJsonValidator], async (req, res) => {
  let start = new Date()
  const buffer = await capture(_.defaults(req.body, { url : kanoUrl, jwt: kanoJwt, delay }))
  if (Buffer.isBuffer(buffer)) {
    res.contentType('image/png')
    res.send(buffer)
    let duration = new Date() - start
    console.log('<> capture processed in %dms', duration)
  } else {
    res.status(500).json({ message: 'Internal service error' })
  }
})

// Healthcheck
app.get('/healthcheck', (req, res) => {
  res.set('Content-Type', 'application/json')
  return res.status(200).json({ isRunning: true })
})

// Serve the app
app.listen(port, () => {
  console.log('[KAPTURE] server listening at %d (body limit %s, delay %s)', port, bodyLimit, delay)
})

